8323546: Cleanup jcmd docs for Compiler.perfmap and VM.cds filename parameter

plummercj opened the pull request:
The jcmd docs for Compiler.perfmap currently say:

    - *filename*: (Optional) The name of the map file (STRING, no default value)

However, there is a default, so not only should that be made more clear in the above, but also some descriptive text as to how the default is generated should be added.

VM.cds has a similar issue, but already has the descriptive text, so just the "no default value" part needs to be fixed.

Another change needed is to consistently use *filename* (italics) instead of `filename` (monospace). Note this is how html formatting is done. For the man page formatting, *filename* does no formatting and `filename` is displayed in color if supported. Personally I prefer `filename`, but it seems that there is already a strong precedence for using italics in the *arguments* list. For example:

    *arguments*:

    - *flag name*: The name of the flag that you want to set (STRING, no
        default value)

    - *string value*: (Optional) The value that you want to set (STRING, no
        default value)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323546](https://bugs.openjdk.org/browse/JDK-8323546): Cleanup jcmd docs for Compiler.perfmap and VM.cds filename parameter (**Enhancement** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [6a9b43f2](https://git.openjdk.org/jdk/pull/17359/files/6a9b43f245e5beddc49c4d944755c08c598fee3a)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17359/head:pull/17359` \
`$ git checkout pull/17359`

Update a local copy of the PR: \
`$ git checkout pull/17359` \
`$ git pull https://git.openjdk.org/jdk.git pull/17359/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17359`

View PR using the GUI difftool: \
`$ git pr show -t 17359`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17359.diff">https://git.openjdk.org/jdk/pull/17359.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17359#issuecomment-1885790338)

plummercj commented:
/label serviceability
/label runtime

plummercj commented:
/label hotspot-runtime

dholmes-ora reviewed (changes requested):
@plummercj the manpage output should be consistent with the help output and AFAICS the help does not state that there is a default filename either for `Compiler.perfmap` or `VM.cds`. I think the help text has to be updated as well, in which case the VM.cds change should be split out separately IMO.

plummercj commented:
I can update the help text, but I don't see why VM.cds needs to be a separate PR.

sspitsyn reviewed (approved):
This looks okay to me modulo the comment from David.

plummercj commented:
Ignore my last push to update the jcmd help output. It doesn't work because it actually sets the default to the specified value, so in the case of `Compiler.perfmap`, it writes to the file `/tmp/perf-<pid>.map` where `<pid>` is literal and not expanded to the actual pid.

I don't think it is going to be possible to get the help output to display the default value since the default value varies based on the pid, and in the case of VM.cds, also depends on the `subcmd` argument.

plummercj commented:
I found a solution. I compare the filename to the default (the descriptive default, not the actual generated default), and if it is the same then the filename is changed to nullptr, which triggers the code that uses it to generate the default file name.

@dholmes-ora @iklam Can you please have a look?

sspitsyn commented on src/hotspot/share/services/diagnosticCommand.cpp:
```diff
@@ -851,15 +851,21 @@ void CodeCacheDCmd::execute(DCmdSource source, TRAPS) {
-  CodeCache::write_perf_map(_filename.value());
+  const char* filename = _filename.value();
+  if (strncmp(filename, DEFAULT_PERFMAP_FILENAME, strlen(DEFAULT_PERFMAP_FILENAME)) == 0) {
+      filename = nullptr; // use the default filename based on the pid
```
Nit: the indent has to be 2, not 4.

dholmes-ora replied:
I only expected to see a change to the help string, I don't understand why the actual Dcmd is being modified here! ???

plummercj replied:
The default value for the argument is what gets displayed in the help text. For example:

```
ThreadDumpToFileDCmd::ThreadDumpToFileDCmd(outputStream* output, bool heap) :
                                           DCmdWithParser(output, heap),
  _overwrite("-overwrite", "May overwrite existing file", "BOOLEAN", false, "false"),
  _format("-format", "Output format (\"plain\" or \"json\")", "STRING", false, "plain"),
  _filepath("filepath", "The file path to the output file", "STRING", true) {
  _dcmdparser.add_dcmd_option(&_overwrite);
  _dcmdparser.add_dcmd_option(&_format);
  _dcmdparser.add_dcmd_argument(&_filepath);
}
```

And the help text:

```
Options: (options must be specified using the <key> or <key>=<value> syntax)
	-overwrite : [optional] May overwrite existing file (BOOLEAN, false)
	-format : [optional] Output format ("plain" or "json") (STRING, plain)
```

The help output that indicates that "plain" is the default format comes from the intialization of the _format argument. There is no separate help text.

dholmes-ora replied:
Ugghh! So the help text is an actual stringification of the actual default value of the "field", whereas in this case the real default value comes from passing null to `CodeCache::write_perf_map`. So we need this hack to deal with that. That is truly awful IMO. The only way to cleanly address that is to expand things so that you can set an actual help string to be used.

iklam replied:
This check is problematic:

```
if (strncmp(filename, DEFAULT_PERFMAP_FILENAME, strlen(DEFAULT_PERFMAP_FILENAME)) == 0)
```

Because it cannot tell whether `filename` was explicitly given by the user. As a result, if the user specifies:

```
jcmd 1234 perfmap '/tmp/perf-<pid>.map'
```

the file will be written as `/tmp/perf-1234.map`, but if the user specifies

```
jcmd 1234 perfmap '/tmp/<pid>-perf.map'
```

then the file will be written as  `/tmp/<pid>-perf.map`.

This gives the confusing impression that the string `<pid>` is sometimes substituted and sometimes not.

I think it's better to do this (similarly for the `VM.cds` case):

```
void PerfMapDCmd::execute(DCmdSource source, TRAPS) {
  CodeCache::write_perf_map(_filename.is_set() ? _filename.value() : nullptr);
}
```

This essentially makes the JVM behavior exactly the same as before, so I think we can change the JBS issue title to something like "Clarify docs for filename parameter to Compiler.perfmap and VM.cds".

dholmes-ora replied:
@iklam but IIUC that does not address the issue of updationg the help output.

iklam replied:
My suggestion is to be used in combination with this change:

```
_filename("filename", "Name of the map file", "STRING", false, DEFAULT_PERFMAP_FILENAME)
```

So the help message will be printed as: 

```
*filename*: (Optional) The name of the map file (STRING /tmp/perf-<pid>.map)
```

dholmes-ora replied:
But that only works if `init_value` is never called, else `is_set` will be true. This seems a rather fragile way to have the "default string" available for the help command, but avoiding actually using it as a default value.

That said, after having a much deeper look at how this all hangs together, it may be this will have to suffice, at least for now. Really we need a way to express that the default-string is only intended for providing the help information it is not actually intended to be parsed to "generate" the default value. But I will also note that part of the problem here is that the DCmd help needs to report the default filename, but the DCmd is not actually responsible for establishing what that default is - we have to hard-wire into the DCMd help knowledge of what the underlying API does.

plummercj replied:
For the dcmd help output, we could take a dynamic approach and have the dcmd call an API (which does not currently exist) which is responsible for generating the default file name. For CDS, this would have to be an upcall to java since that is where the default filename is currently generated.

...or we could just punt on this for now and continue with not actually specifying the default in the help output or the man page.

dholmes-ora replied:
I don't think we want that for the help output - I think we want the generic form with the "<pid>" text, not the filename with the actual pid value inserted.

I do think we want the help and manpage to document the default filenames. So I think Ioi's suggestion is the best option at the moment. Thanks.

dholmes-ora reviewed (commented):
I must not understand how the help text is generated because I did not expect to see the changes to the functional code.

And I still think the VM.CDS change show have its own issue (if not PR).

plummercj commented:
> And I still think the VM.CDS change show have its own issue (if not PR).

Yes, and I asked why that is the case. The VM.cds changes are very much related to the Compiler.perfmap changes. I think it's actually correct to do them together, not just easier.

dholmes-ora commented:
> > And I still think the VM.CDS change show have its own issue (if not PR).
> 
> Yes, and I asked why that is the case. The VM.cds changes are very much related to the Compiler.perfmap changes. I think it's actually correct to do them together, not just easier.

And I gave an answer. At a minimum if you insist of doing these two distinct updates together then the JBS issue needs a new title and description.

plummercj commented:
> > > And I still think the VM.CDS change show have its own issue (if not PR).
> > 
> > 
> > Yes, and I asked why that is the case. The VM.cds changes are very much related to the Compiler.perfmap changes. I think it's actually correct to do them together, not just easier.
> 
> And I gave an answer. At a minimum if you insist of doing these two distinct updates together then the JBS issue needs a new title and description.

Sorry I missed your response, although I still honestly don't know when/where it was. I looked throughout this PR and the CR and don't see any suggestion to update the synopsis, but certainly that is needed. I'll take care of it once we've settled the other issues in this PR.

dholmes-ora commented:
> Sorry I missed your response, although I still honestly don't know when/where it was. 

No I'm sorry, I can't find it either. I must have forgotten to click the right button at some point. :( My concern was that once we were changing functional code (not just the manpage) then I didn't want a functional change to the CDS DCmd code hidden in this issue doc issue.

> I looked throughout this PR and the CR and don't see any suggestion to update the synopsis, but certainly that is needed. 

No that was a new suggestion for if you decided to keep them together.

plummercj commented:
I've updated the PR with Ioi's suggestion and also added a test cast to PerfMapTest.java to make sure if the default is specified, then it actually gets used as the filename (it failed before I added Ioi's fix). Ioi, let me know if you think a similar test for VM.cds is warranted.

i also updated the javadoc to not include quotes around the default filename. This is consistent with how other defaults are printed, and also how jcmd "help" prints the default argument.

iklam reviewed (approved):
LGTM. I think having one test case for perfmap should be enough.

dholmes-ora reviewed (approved):
Looks good.

Thanks for all the effort on this one.

plummercj commented:
Thanks for the reviews David, Ioi, and Serguei!

/integrate
